Kernel/Net/Intel/E1000NetworkAdapter.cpp

```diff
@@ -207,35 +304,127 @@ UNMAP_AFTER_INIT ErrorOr<void> E1000NetworkAdapter::initialize(Badge<NetworkingM
+        return;
+    }
+    // 8254x \(82541xx, 82547GI/EI)
+    if (first_is_one_of(device_id,
+            0x100E, // 82540EM-A
+            0x100F, // 82545EM-A (COPPER)
+            0x1010, // 82546EB-A1 (COPPER)
+            0x1011, // 82545EM-A (FIBER)
+            0x1012, // 82546EB-A1 (FIBER)
+            0x1015, // 82540EM-A (LOM)
+            0x1016, // 82540EP-A (LOM)
+            0x1017, // 82540EP-A
+            0x1019, // 82547EI-A0, 82547EI-A1, 82547EI-B0, 82547GI-B0
+            0x101A, // 82547EI-B0 (MOBILE)
+            0x101D, // 82546EB-A1 (QUAD-COPPER)
+            0x1026, // 82545GM-B (COPPER)
+            0x1027, // 82545GM-B (FIBER)
+            0x1028, // 82545GM-B (SERDES)
+            0x1079, // 82546GB-B0 (COPPER)
+            0x107A, // 82546GB-B0 (FIBER)
+            0x107B  // 82546GB-B0 (SERDES)
+            )) {
+
+        dbgln("E1000: Using Intel8254x mode");
+        m_operating_mode = OperatingMode::Intel8254x;
+
+        return;
+    }
+    // 82541xx, 82547GI/EI
+    // 63[12]xESB
+    // 8256[34]EB
+    // 8257[123]
+    // 82574
+    if (first_is_one_of(device_id,
+            0x1013, // 82541EI
+            0x1013, // 82541EI-A0, 82541EI-B0
+            0x1014, // 82541ER_LOM
+            0x1018, // 82541EI-B0
+            0x1019, // 82547EI-A0, 82547EI-A1, 82547EI-B0, 82547GI-B0
+            0x101A, // 82547EI-B0
+            0x1076, // 82541GI
+            0x1076, // 82541GI-B1, 82541PI-C0
+            0x1077, // 82541GI_MOBILE
+            0x1077, // 82541GI-B1
+            0x1078, // 82541ER
+            0x1078, // 82541ER-C0
+            0x107C, // 82541GI_LF
+            // 63[12]xESB
+            // 8256[34]EB
+            // 8257[123]
+            0x105E,
+            0x1081,
+            0x1082,
+            0x1083,
+            0x1096,
+            0x1097,
+            0x1098,
+            0x108B,
+            0x108C,
+
+            0x10D3 // 82574L
+            )) {
+        dbgln("E1000: Late E1000 mode");
+
+        m_operating_mode = OperatingMode::Intel8254x_14bit_til_82574;
+        return;
+    }
+    dbgln("E1000: Usign IGB/82576 and later mode");
+    // 82576 and later
+    m_operating_mode = OperatingMode::Intel82576_and_later;
+}
+
 UNMAP_AFTER_INIT void E1000NetworkAdapter::setup_link()
 {
-    u32 flags = in32(REG_CTRL);
-    out32(REG_CTRL, flags | ECTRL_SLU);
+    auto ctrl = m_registers.read<Register::Ctrl>();
+    ctrl.set_link_up = 1;
+    m_registers.write<Register::Ctrl>(ctrl);
 }
 
 UNMAP_AFTER_INIT void E1000NetworkAdapter::setup_interrupts()
 {
-    out32(REG_INTERRUPT_RATE, 6000); // Interrupt rate of 1.536 milliseconds
-    out32(REG_INTERRUPT_MASK_SET, INTERRUPT_LSC | INTERRUPT_RXT0 | INTERRUPT_RXO);
-    in32(REG_INTERRUPT_CAUSE_READ);
+    // FIXME: Do this properly,
+    //        There's probably a way to find an ideal interrupt rate etc
+    m_registers.write<Register::InterruptThrottling>(6000); // Interrupt rate of 1.536 milliseconds
```

Comment:
We should be able to calculate the ideal rate dynamically based on packet/byte counts during the last firing of the interrupt. We have to test what the ideal boundaries are - so it may be out of scope for this PR. Would be great though if there is some sort of function that runs after each interrupt though if we decide to go forward with this.